### PR TITLE
Prepare flutter engine for enabling private final field promotion.

### PR DESCRIPTION
**Pull request by stereotype441**
Some parts of the flutter engine are built using the latest version of the Dart SDK, ignoring SDK constraints in `pubspec.yaml` files. Therefore, before the Dart SDK can switch on the "private field promotion" feature (https://github.com/dart-lang/language/issues/2020), these parts of the flutter engine need to be modified so that they won't have any "unnecessary `!`" warnings after field private field promotion is enabled. This PR makes the necessary changes.

This PR doesn't introduce any functional change. In principle it might improve performance slightly (by avoiding redundant memory accesses), but in practice the difference is unlikely to rise out of the measurement noise.

Issue fixed by this PR: https://github.com/flutter/flutter/issues/131198

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Review by goderbauer — APPROVED** [review]
LGTM

File: lib/web_ui/lib/src/engine/semantics/semantics.dart [hunk 1]
```diff
@@ -2122,14 +2122,15 @@ class EngineSemanticsOwner {
     // First, update each object's information about itself. This information is
     // later used to fix the parent-child and sibling relationships between
     // objects.
-    for (final SemanticsNodeUpdate nodeUpdate in update._nodeUpdates!) {
+    final List<SemanticsNodeUpdate> nodeUpdates = update._nodeUpdates!;
```
Comment (Hixie) [inline lib/web_ui/lib/src/engine/semantics/semantics.dart:2125]: please add an assert where the `nodeUpdates` variable goes out of scope that verifies that `update._nodeUpdates` has not changed

**Comment by Hixie** [comment]
test-exempt: code refactor with no semantic change

File: lib/web_ui/lib/src/engine/semantics/semantics.dart (hunk 1 above)
Comment (stereotype441) [reply lib/web_ui/lib/src/engine/semantics/semantics.dart:2125]: Done

**Comment by goderbauer** [comment]
@jonahwilliams (engine sherif) do you know what's up with the "Linux Fuchsia FEMU" check? It keeps timing out.

**Comment by jonahwilliams** [comment]
I do not. I'm going to click rebase just in case, but I don't see this happening elsewhere so hopefully its just a flake...

**Comment by stereotype441** [comment]
> I do not. I'm going to click rebase just in case, but I don't see this happening elsewhere so hopefully its just a flake...

Thanks, Jonah! It looks like everything is working now.

**Pull request merged**